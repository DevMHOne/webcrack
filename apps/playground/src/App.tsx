import * as monaco from 'monaco-editor';
import {
  For,
  Show,
  batch,
  createEffect,
  createMemo,
  createSignal,
  onCleanup,
} from 'solid-js';
import { createStore } from 'solid-js/store';
import Breadcrumbs from './components/Breadcrumbs';
import MonacoEditor from './components/MonacoEditor';
import Sidebar from './components/Sidebar';
import Tab from './components/Tab';
import { DeobfuscateContextProvider } from './context/DeobfuscateContext';
import {
  clearSavedModels,
  loadSavedModels,
  saveModels,
  type SavedModel,
} from './indexeddb';
import type { DeobfuscateResult } from './webcrack.worker';

export const [settings, setSettings] = createStore({
  deobfuscate: true,
  unminify: true,
  unpack: true,
  jsx: true,
  mangle: false,
});

function App() {
  const [untitledCounter, setUntitledCounter] = createSignal(1);
  const [models, setModels] = createSignal<monaco.editor.ITextModel[]>([
    monaco.editor.createModel(
      '',
      'javascript',
      monaco.Uri.from({ scheme: 'untitled', path: 'Untitled-1' }),
    ),
  ]);
  const [tabs, setTabs] = createSignal<monaco.editor.ITextModel[]>(models());
  const [activeTab, setActiveTab] = createSignal<
    monaco.editor.ITextModel | undefined
  >(tabs()[0]);
  const [savedModels, setSavedModels] = createSignal<SavedModel[]>([]);

  const fileModels = createMemo(() =>
    models().filter((m) => m.uri.scheme === 'file'),
  );
  const untitledModels = createMemo(() =>
    models().filter((m) => m.uri.scheme === 'untitled'),
  );
  const filePaths = createMemo(() =>
    fileModels().map((model) => model.uri.path),
  );

  loadSavedModels().then(setSavedModels).catch(console.error);
  setTimeout(() => setSavedModels([]), 5000);

  createEffect(() => {
    if (
      models().length === 0 ||
      models().every((m) => m.getValueLength() === 0)
    ) {
      return;
    }
    saveModels(models()).catch(console.error);
  });

  function restoreSavedModels() {
    batch(() => {
      models().forEach((model) => model.dispose());

      setModels(
        savedModels().map((model) =>
          monaco.editor.createModel(
            model.value,
            model.language,
            monaco.Uri.parse(model.uri),
          ),
        ),
      );
      console.log(
        'Loaded saved models',
        savedModels().map((m) => m.uri),
      );
      setSavedModels([]);

      setTabs(untitledModels());
      setActiveTab(untitledModels()[0]);
    });
  }

  onCleanup(() => {
    models().forEach((model) => model.dispose());
  });

  function openTab(tab: monaco.editor.ITextModel) {
    if (!tabs().includes(tab)) {
      setTabs([...tabs(), tab]);
    }
    setActiveTab(tab);
  }

  function openFile(path: string) {
    const model = fileModels().find((m) => m.uri.path === '/' + path);
    if (!model) {
      return console.warn(`No model found for path: ${path}`);
    }
    openTab(model);
  }

  function closeTab(tab: monaco.editor.ITextModel) {
    const index = tabs().indexOf(tab);
    if (activeTab() === tab) {
      setActiveTab(tabs()[index > 0 ? index - 1 : 1]);
    }
    setTabs(tabs().filter((t) => t !== tab));
    if (tab.uri.scheme === 'untitled') {
      tab.dispose();
      // FIXME: resets folder expansion state
      setModels(models().filter((m) => m !== tab));
    }
  }

  function openUntitledTab() {
    setUntitledCounter(untitledCounter() + 1);
    const model = monaco.editor.createModel(
      '',
      'javascript',
      monaco.Uri.from({
        scheme: 'untitled',
        path: `Untitled-${untitledCounter()}`,
      }),
    );
    setModels([...models(), model]);
    openTab(model);
    return model;
  }

  function onDeobfuscateResult(result: DeobfuscateResult) {
    let model = activeTab();

    if (result.files.length === 0) {
      model ||= openUntitledTab();
      model.pushEditOperations(
        [],
        [
          {
            range: model.getFullModelRange(),
            text: result.code,
          },
        ],
        () => null,
      );
      return;
    }

    if (!model || model.uri.scheme === 'file') {
      model = openUntitledTab();
    }

    model.setValue(result.code);

    fileModels().forEach((model) => model.dispose());
    setTabs(untitledModels());

    setModels([
      ...untitledModels(),
      ...result.files.map((file) =>
        monaco.editor.createModel(
          file.code,
          'javascript',
          monaco.Uri.file(file.path),
        ),
      ),
      monaco.editor.createModel(
        result.code,
        'javascript',
        monaco.Uri.file('deobfuscated.js'),
      ),
    ]);
  }

  function onDeobfuscateError(error: unknown) {
    console.error(error);
  }

  return (
    <DeobfuscateContextProvider
      code={activeTab()?.getValue()}
      options={{ ...settings }}
      onResult={onDeobfuscateResult}
      onError={onDeobfuscateError}
    >
      <Show when={savedModels().length > 0}>
        <div role="alert" class="alert absolute z-10 bottom-5 right-5 max-w-md">
          <svg
            xmlns="http://www.w3.org/2000/svg"
            fill="none"
            viewBox="0 0 24 24"
            class="stroke-info shrink-0 w-6 h-6"
          >
            <path
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              d="M13 16h-1v-4h-1m1-4h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
            ></path>
          </svg>
          <span>
            Restore {savedModels().length} saved tabs from last session?
          </span>
          <div>
            <button
              class="btn btn-sm"
              onClick={() => {
                void clearSavedModels();
                setSavedModels([]);
              }}
            >
              Clear
            </button>
            <button class="btn btn-sm btn-primary" onClick={restoreSavedModels}>
              Load
            </button>
          </div>
        </div>
      </Show>

      <div class="h-screen flex">
        <Sidebar paths={filePaths()} onFileClick={openFile} />

        <main class="flex-1 overflow-auto">
          <div class="tabs tabs-lifted justify-start overflow-x-auto bg-base-300">
            <For each={tabs()}>
              {(tab) => (
                <Tab
                  path={tab.uri.path}
                  active={activeTab() === tab}
                  onClick={() => setActiveTab(tab)}
                  onClose={() => closeTab(tab)}
                />
              )}
            </For>
            <div class="tab" title="New tab" onClick={openUntitledTab}>
              <svg
                width="16"
                height="16"
                viewBox="0 0 24 24"
                stroke-width="1.5"
                stroke="currentColor"
                fill="none"
                stroke-linecap="round"
                stroke-linejoin="round"
              >
                <path stroke="none" d="M0 0h24v24H0z" fill="none" />
                <path d="M12 5l0 14" />
                <path d="M5 12l14 0" />
              </svg>
            </div>
          </div>
          <Show when={activeTab()?.uri.scheme === 'file'}>
            <Breadcrumbs path={activeTab()!.uri.path} />
          </Show>

          <MonacoEditor
            models={models()}
            currentModel={activeTab()}
            onModelChange={openTab}
            onValueChange={() => {
              saveModels(models()).catch(console.error);
            }}
          />
        </main>
      </div>
    </DeobfuscateContextProvider>
  );
}

export default App;
