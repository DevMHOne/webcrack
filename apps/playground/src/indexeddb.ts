import { openDB, type DBSchema } from 'idb';
import type * as monaco from 'monaco-editor';

export interface SavedModel {
  value: string;
  language: string;
  uri: string;
}

interface MyDB extends DBSchema {
  models: {
    key: string;
    value: SavedModel;
  };
}

async function initDB() {
  return openDB<MyDB>('models', 1, {
    upgrade(db) {
      db.createObjectStore('models', { keyPath: 'uri' });
    },
  });
}

export async function saveModels(models: monaco.editor.ITextModel[]) {
  console.log('Saving models...', models.length);
  const db = await initDB();
  await db.clear('models');
  await Promise.all(
    models.map(async (model) => {
      await db.put('models', {
        value: model.getValue(),
        language: model.getLanguageId(),
        uri: model.uri.toString(),
      });
    }),
  );
}

export async function clearSavedModels() {
  const db = await initDB();
  await db.clear('models');
}

export async function saveModel(model: monaco.editor.ITextModel) {
  const db = await initDB();
  await db.put('models', {
    value: model.getValue(),
    language: model.getLanguageId(),
    uri: model.uri.toString(),
  });
}

export async function loadSavedModels(): Promise<SavedModel[]> {
  const db = await initDB();
  return db.getAll('models');
}
