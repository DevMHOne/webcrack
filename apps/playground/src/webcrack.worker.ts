import type { Options, Sandbox} from 'webcrack';
import { webcrack } from 'webcrack';

export type WorkerRequest =
  | { type: 'deobfuscate'; code: string; options: Options }
  | { type: 'sandbox'; result: unknown };

export type WorkerResponse =
  | { type: 'sandbox'; code: string }
  | ({ type: 'result' } & DeobfuscateResult)
  | { type: 'progress'; value: number }
  | { type: 'error'; error: unknown };

export interface DeobfuscateResult {
  code: string;
  files: { code: string; path: string }[];
}

const postMessage = (message: WorkerResponse) => self.postMessage(message);

self.onmessage = async ({ data }: MessageEvent<WorkerRequest>) => {
  if (data.type !== 'deobfuscate') return;

  // worker->window->sandybox because it accesses the DOM, which is not available in workers
  const sandbox: Sandbox = (code) => {
    return new Promise((resolve) => {
      self.addEventListener('message', onSandboxResponse);
      postMessage({ type: 'sandbox', code });

      function onSandboxResponse({ data }: MessageEvent<WorkerRequest>) {
        if (data.type === 'sandbox') {
          self.removeEventListener('message', onSandboxResponse);
          resolve(data.result);
        }
      }
    });
  };

  function onProgress(value: number) {
    postMessage({ type: 'progress', value });
  }

  try {
    const result = await webcrack(data.code, {
      sandbox,
      onProgress,
      ...data.options,
    });
    const files = Array.from(result.bundle?.modules ?? [], ([, module]) => ({
      code: module.code,
      path: module.path.replace(/\.?\/?/, ''),
    }));

    postMessage({ type: 'result', code: result.code, files });
  } catch (error) {
    postMessage({ type: 'error', error });
  }
};
