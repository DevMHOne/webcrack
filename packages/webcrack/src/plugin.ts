import { parse } from '@babel/parser';
import template from '@babel/template';
import traverse, { visitors, type Visitor } from '@babel/traverse';
import * as t from '@babel/types';
import * as m from '@codemod/matchers';

const stages = [
  'parse',
  'prepare',
  'deobfuscate',
  'unminify',
  'unpack',
] as const;

export type Stage = (typeof stages)[number];

export type PluginState = { opts: Record<string, unknown> };

export interface PluginObject {
  name?: string;
  runAfter: Stage;
  pre?: (this: PluginState, state: PluginState) => Promise<void> | void;
  post?: (this: PluginState, state: PluginState) => Promise<void> | void;
  visitor?: Visitor<PluginState>;
}

export interface PluginAPI {
  parse: typeof parse;
  types: typeof t;
  traverse: typeof traverse;
  template: typeof template;
  matchers: typeof m;
}

export type Plugin = (api: PluginAPI) => PluginObject;

export function loadPlugins(plugins: Plugin[]) {
  const groups = new Map<Stage, PluginObject[]>(
    stages.map((stage) => [stage, []]),
  );
  for (const plugin of plugins) {
    const obj = plugin({
      parse,
      types: t,
      traverse,
      template,
      matchers: m,
    });
    groups.get(obj.runAfter)?.push(obj);
  }
  return Object.fromEntries(
    [...groups].map(([stage, plugins]) => [
      stage,
      plugins.length
        ? async (ast: t.File) => {
            const state: PluginState = { opts: {} };
            for (const transform of plugins) {
              await transform.pre?.call(state, state);
            }

            const pluginVisitors = plugins.flatMap(
              (plugin) => plugin.visitor ?? [],
            );
            if (pluginVisitors.length > 0) {
              const mergedVisitor = visitors.merge(pluginVisitors);
              traverse(ast, mergedVisitor, undefined, state);
            }

            for (const plugin of plugins) {
              await plugin.post?.call(state, state);
            }
          }
        : undefined,
    ]),
  ) as Record<Stage, (ast: t.File) => Promise<void>>;
}
