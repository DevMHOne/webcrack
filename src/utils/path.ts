import { dirname, relative } from 'node:path/posix';

export function relativePath(from: string, to: string) {
  const relativePath = relative(dirname(from), to);
  return relativePath.startsWith('.') ? relativePath : './' + relativePath;
}

/**
 * Resolve the path of each module of a browserify bundle
 * based on its dependencies.
 * @param tree module id -> dependencies (id -> path)
 * @param entry entry module id
 */
export function resolveDependencyTree(
  tree: Record<number, Record<number, string>>,
  entry: number
): Record<number, string> {
  const paths = resolveTreePaths(tree, entry);
  paths[entry] = './index.js';

  const entryDepth = Object.values(paths).reduce(
    (acc, path) => Math.max(acc, path.split('..').length),
    0
  );
  // If the entrypoint is in a subfolder, we need to make up a prefix to get rid of the `../`
  const prefix = Array(entryDepth - 1)
    .fill(0)
    .map((_, i) => `tmp${i}`)
    .join('/');

  return Object.fromEntries(
    Object.entries(paths).map(([id, path]) => {
      const newPath = path.startsWith('node_modules/')
        ? path
        : join(prefix, path);
      assert(!newPath.includes('..'));
      return [id, newPath];
    })
  );
}

/**
 * Recursively resolve the paths of a dependency tree.
 */
function resolveTreePaths(
  graph: Record<number, Record<number, string>>,
  entry: number,
  cwd = '.'
) {
  const paths: Record<number, string> = {};
  const entries = Object.entries(graph[entry]) as unknown as [number, string][];

  for (const [id, key] of entries) {
    let path: string;
    if (key.startsWith('.')) {
      path = join(cwd, key);
      if (!path.endsWith('.js')) path += '.js';
    } else {
      path = join('node_modules', key, 'index.js');
    }
    paths[id] = path;
    const newCwd = path.endsWith('.js') ? dirname(path) : path;
    Object.assign(paths, resolveTreePaths(graph, id, newCwd));
  }

  return paths;
}
