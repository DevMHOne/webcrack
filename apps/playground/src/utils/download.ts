import type * as monaco from 'monaco-editor';
import { basename } from 'path';

export function downloadFile(model: monaco.editor.ITextModel) {
  const blob = new Blob([model.getValue()], {
    type: 'application/javascript;charset=utf-8',
  });
  const url = URL.createObjectURL(blob);
  const link = document.createElement('a');

  link.setAttribute('href', url);
  link.setAttribute('download', basename(model.uri.fsPath));
  link.style.display = 'none';
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);

  URL.revokeObjectURL(url);
}
