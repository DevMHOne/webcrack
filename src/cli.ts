#!/usr/bin/env node

import { program } from 'commander';
import { existsSync, readFileSync } from 'node:fs';
import { readFile, rm } from 'node:fs/promises';
import { join } from 'node:path';
import * as url from 'url';
import { webcrack } from '.';

const __dirname = url.fileURLToPath(new URL('.', import.meta.url));

const { version, description } = JSON.parse(
  readFileSync(join(__dirname, '..', 'package.json'), 'utf8')
);

program
  .version(version)
  .description(description)
  .option('-o, --output <path>', 'output directory', 'webcrack-out')
  .option('-f, --force', 'overwrite output directory')
  .argument('<file>', 'input file')
  .action(async input => {
    const { output, force } = program.opts();
    const code = await readFile(input, 'utf8');

    if (force || !existsSync(output)) {
      await rm(output, { recursive: true, force: true });
    } else {
      program.error('output directory already exists');
    }

    (await webcrack(code)).save(output);
  })
  .parse();
