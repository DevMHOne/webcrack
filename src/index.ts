import generate from '@babel/generator';
import { parse } from '@babel/parser';
import * as m from '@codemod/matchers';
import { join } from 'node:path';
import deobfuscator from './deobfuscator';
import debugProtection from './deobfuscator/debugProtection';
import selfDefending from './deobfuscator/selfDefending';
import { Bundle, extractBundle } from './extractor';
import { applyTransform } from './transforms';
import jsx from './transforms/jsx';
import unminify from './transforms/unminify';

export interface WebcrackResult {
  code: string;
  bundle: Bundle | undefined;
  /**
   * Save the deobufscated code and the extracted bundle to the given directory.
   * @param path Output directory
   */
  save(path: string): Promise<void>;
}

export interface Options {
  /**
   * Run for every module after generating the code and before saving it.
   * This can be used to format the code or apply other transformations.
   */
  transformCode?: (code: string) => Promise<string> | string;
  /**
   * Assigns paths to modules based on the given matchers.
   * This will also rewrite `require()` calls to use the new paths.
   *
   * @example
   * ```js
   * m => ({
   *   './utils/color.js': m.regExpLiteral('^#([0-9a-f]{3}){1,2}$')
   * })
   * ```
   */
  mappings?: (
    m: typeof import('@codemod/matchers')
  ) => Record<string, m.Matcher<unknown>>;
}

export async function webcrack(
  code: string,
  options: Options = {}
): Promise<WebcrackResult> {
  const ast = parse(code, {
    sourceType: 'unambiguous',
    allowReturnOutsideFunction: true,
  });

  applyTransform(ast, deobfuscator);
  applyTransform(ast, unminify);

  // Have to run this after readability transforms because the function may contain dead code
  applyTransform(ast, selfDefending);
  applyTransform(ast, debugProtection);

  applyTransform(ast, jsx);

  const bundle = extractBundle(ast);
  console.log('Bundle:', bundle?.type);

  let outputCode = generate(ast).code;
  outputCode = options.transformCode
    ? await options.transformCode(outputCode)
    : outputCode;

  return {
    code: outputCode,
    bundle,
    async save(path) {
      if (process.env.browser) {
        throw new Error('Not implemented.');
      } else {
        const { mkdir, writeFile } = await import('node:fs/promises');
        await mkdir(path, { recursive: true });
        await writeFile(join(path, 'deobfuscated.js'), outputCode, 'utf8');
        bundle?.save(path, options.transformCode, options.mappings);
      }
    },
  };
}
