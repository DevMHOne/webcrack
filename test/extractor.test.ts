import * as m from '@codemod/matchers';
import assert from 'assert';
import { readFile } from 'fs/promises';
import { join } from 'path';
import { describe, expect, test } from 'vitest';
import { webcrack } from '../src/index';
import { relativePath } from '../src/utils/path';

// Test samples
test.each([
  'webpack.js',
  'webpack_object.js',
  'webpack-esm.js',
  'webpack-var-injection.js',
  'webpack5_object.js',
])('extract %s', async filename => {
  const { bundle } = await webcrack(
    await readFile(join('test', 'samples', filename), 'utf8')
  );
  assert(bundle);
  bundle.applyTransforms();
  for (const module of bundle.modules.values()) {
    expect(module.ast).toMatchSnapshot();
  }
});

describe('extractor', async () => {
  test('webpack array', async () => {
    const { bundle } = await webcrack(
      await readFile('./test/samples/webpack.js', 'utf8')
    );
    expect(bundle).toBeDefined();
    assert(bundle);
    expect(bundle.type).toBe('webpack');
    expect(bundle.entryId).toBe(2);
  });

  test('webpack object', async () => {
    const { bundle } = await webcrack(
      await readFile('./test/samples/webpack_object.js', 'utf8')
    );
    expect(bundle).toBeDefined();
    assert(bundle);
    expect(bundle.type).toBe('webpack');
    expect(bundle.entryId).toBe(386);
  });

  test('path mapping', async () => {
    const { bundle } = await webcrack(
      await readFile('./test/samples/webpack.js', 'utf8')
    );
    expect(bundle).toBeDefined();
    assert(bundle);
    bundle.applyMappings({ './utils/color.js': m.stringLiteral('#FBC02D') });
    bundle.applyTransforms();

    for (const module of bundle.modules.values()) {
      expect(module.ast).toMatchSnapshot();
    }
  });
});

test('relative paths', () => {
  expect(relativePath('./a.js', './x/y.js')).toBe('./x/y.js');
  expect(relativePath('./x/y.js', './a.js')).toBe('../a.js');
});
