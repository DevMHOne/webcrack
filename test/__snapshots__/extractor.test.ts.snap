// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extract webpack.js 1`] = `const a = require("./3.js");`;

exports[`extract webpack.js 2`] = `
const a = require("./1.js");
const _module = 1;
module.exports.color = "#FBC02D";
{
  const _module2 = 2;
  console.log(_module2);
  console.log(module);
}
exports.a = 3;
`;

exports[`extract webpack.js 3`] = `module.exports = 4;`;

exports[`extract webpack_object.js 1`] = `const r = require("./387.js").default;`;

exports[`extract webpack_object.js 2`] = `module.exports = "test";`;

exports[`extract webpack-esm.js 1`] = `
const a = require("./1.js");
const b = require("./2.js");
const c = require("./3.js");
const bDefault = b.default;
const cDefault = c;
const dDefault = b.default;
const eDefault = c;
console.log(a.counter);
console.log(bDefault.VERSION);
console.log(cDefault.VERSION);
`;

exports[`extract webpack-esm.js 2`] = `
export class counter {}
let _counter = 1;
counter = 2;
`;

exports[`extract webpack-esm.js 3`] = `
const x = {};
export default 1;
`;

exports[`extract webpack-esm.js 4`] = `
module.exports = {
  VERSION: 1
};
`;

exports[`extract webpack-var-injection.js 1`] = `
var m = require("./1.js");
var n = require("./2.js");
console.log(m, n);
`;

exports[`extract webpack-var-injection.js 2`] = `module.exports = 1;`;

exports[`extract webpack-var-injection.js 3`] = `module.exports = 2;`;

exports[`extract webpack5_object.js 1`] = `
const lib = require("./3.js");
console.log(lib);
`;

exports[`extract webpack5_object.js 2`] = `module.exports = "foo";`;

exports[`extractor > path mapping 1`] = `const a = require("./3.js");`;

exports[`extractor > path mapping 2`] = `
const a = require("../1.js");
const _module = 1;
module.exports.color = "#FBC02D";
{
  const _module2 = 2;
  console.log(_module2);
  console.log(module);
}
exports.a = 3;
`;

exports[`extractor > path mapping 3`] = `module.exports = 4;`;
